tools/base/fpgad/plugins/fpgad-dc/fpgad-dc.c

```diff
@@ -0,0 +1,937 @@
+		} else {
+			s->high_fatal = s->high_fatal_cfg * FPGA_SENSOR_MILLI;
+			s->high_fatal -= (s->high_fatal * DC_PERCENT_ADJUST_PWR) / 100;
+		}
+
+	 }
+
+	// Set High warn Config Values
+	if ((s->flags & FPGAD_SENSOR_DC_HIGH_WARN_VALID) &&
+		s->high_warn_cfg != DBL_MAX) {
+
+		if (is_temperature) {
+			s->high_warn = s->high_warn_cfg * FPGA_SENSOR_MILLI;
+			s->high_warn -= DC_DEGREES_ADJUST_TEMP;
+		} else {
+			s->high_warn = s->high_warn_cfg * FPGA_SENSOR_MILLI;
+			s->high_warn -= (s->high_warn * DC_PERCENT_ADJUST_PWR) / 100;
+		}
+	}
+
+	// Set low fatal Config Values
+	if ((s->flags & FPGAD_SENSOR_DC_LOW_FATAL_VALID) &&
+		s->low_fatal_cfg != DBL_MAX) {
+
+		if (is_temperature) {
+			s->low_fatal = s->low_fatal_cfg * FPGA_SENSOR_MILLI;
+			s->low_fatal += DC_DEGREES_ADJUST_TEMP;
+		} else {
+			s->low_fatal = s->low_fatal_cfg * FPGA_SENSOR_MILLI;
+			s->low_fatal += (s->low_fatal * DC_PERCENT_ADJUST_PWR) / 100;
+		}
+
+	}
+
+	// Set low warn Config Values
+	if ((s->flags & FPGAD_SENSOR_DC_LOW_WARN_VALID) &&
+		s->low_warn_cfg != DBL_MAX) {
+
+		if (is_temperature) {
+			s->low_warn = s->low_warn_cfg * FPGA_SENSOR_MILLI;
+			s->low_warn -= DC_DEGREES_ADJUST_TEMP;
+		} else {
+			s->low_warn = s->low_warn_cfg * FPGA_SENSOR_MILLI;
+			s->low_warn -= (s->low_warn * DC_PERCENT_ADJUST_PWR) / 100;
+		}
+
+	}
+
+	return FPGA_OK;
+}
+
+STATIC fpga_result dc_enum_sensor(dc_device *dc,
+				  const char *name)
+{
+	fpga_result res;
+	dc_sensor *s;
+
+	if (dc->num_sensors == MAX_DC_SENSORS) {
+		LOG("exceeded max number of sensors.\n");
+		return FPGA_EXCEPTION;
+	}
+
+	s = &dc->sensors[dc->num_sensors];
+
+	res = fpgaObjectGetObject(dc->group_object,
+				  name,
+				  &s->sensor_object,
+				  0);
+
+	if (res != FPGA_OK)
+		return res;
+
+	res = fpgaObjectGetObject(s->sensor_object,
+				  FPGA_SENSOR_VALUE,
+				  &s->value_object,
+				  0);
+
+	if (res != FPGA_OK) {
+		LOG("failed to get value object for %s.\n", name);
+		fpgaDestroyObject(&s->sensor_object);
+		return res;
+	}
+	res = dc_sensor_get(s);
+	if (res == FPGA_OK)
+		++dc->num_sensors;
+	else {
+		LOG("warning: sensor attribute enumeration failed.\n");
+		dc_destroy_sensor(s);
+	}
+
+	return res;
+}
+
+STATIC fpga_result dc_enum_sensors(dc_device *dc)
+{
+	fpga_result res;
+	char name[SYSFS_PATH_MAX];
+	int i;
+
+	res = fpgaTokenGetObject(dc->base_device->token,
+				 FPGA_DC_SENSOR_PATH,
+				 &dc->group_object,
+				 FPGA_OBJECT_GLOB);
+
+	if (res)
+		return res;
+
+	for (i = 0 ; i < MAX_DC_SENSORS ; ++i) {
+		snprintf_s_i(name, sizeof(name), "sensor%d", i);
+		dc_enum_sensor(dc, name);
+	}
+
+	if (dc->num_sensors > 0) {
+		dc_sensor *s = &dc->sensors[dc->num_sensors - 1];
+
+		dc->max_sensor_id = s->id;
+
+		dc->state_tripped = calloc((dc->max_sensor_id + 7) / 8, 1);
+		dc->state_last = calloc((dc->max_sensor_id + 7) / 8, 1);
+
+		return (dc->state_tripped && dc->state_last) ?
+			FPGA_OK : FPGA_NO_MEMORY;
+	}
+
+	return FPGA_NOT_FOUND;
+}
+
+STATIC int tripped_count;
```

Comment:
I've since made tripped_count a member of the vc_device.  Should probably do the same here.